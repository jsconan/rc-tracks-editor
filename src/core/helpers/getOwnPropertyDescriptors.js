/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Gets a list of descriptors to copy from a source object to a target object.
 * If the target object is given, the properties that already exist in the target,
 * or in the listed descriptors, will not be copied.
 * @param {object} source - The source object where to extract the property descriptors.
 * @param {object} descriptors - A descriptors map to build.
 * @param {Array} keys - A list of property names to get the descriptor for.
 * @param {object} target - The target object that is expected to receive the properties.
 * @returns {number} - The number of property descriptors that have been added to the descriptors map.
 */
export default (source, descriptors, keys, target = null) => {
    const hasOwn = Object.prototype.hasOwnProperty;
    const len = keys.length;
    let count = 0;
    let i = 0;

    while (i < len) {
        const key = keys[i++];
        if (!target || (!descriptors[key] && !hasOwn.call(target, key))) {
            const descriptor = Object.getOwnPropertyDescriptor(source, key);
            if ('undefined' !== typeof descriptor) {
                descriptors[key] = descriptor;
                count++;
            }
        }
    }

    return count;
};
