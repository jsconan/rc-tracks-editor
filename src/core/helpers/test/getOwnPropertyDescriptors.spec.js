/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import getOwnPropertyDescriptors from '../getOwnPropertyDescriptors.js';

describe('getOwnPropertyDescriptors', () => {
    it('is a function', () => {
        expect(getOwnPropertyDescriptors).toEqual(expect.any(Function));
    });

    describe('lists descriptors', () => {
        it('when the source contains the expected property', () => {
            const source = { foo: 'bar', bar: 'foo' };
            const descriptors = {};
            const keys = ['foo'];
            expect(getOwnPropertyDescriptors(source, descriptors, keys)).toBe(1);
            expect(descriptors).toStrictEqual({
                foo: {
                    configurable: true,
                    enumerable: true,
                    writable: true,
                    value: 'bar'
                }
            });
        });

        it.each([
            ['string', 'message', { message: 'hello' }],
            ['boolean', 'flag', { flag: true }],
            ['number', 'answer', { answer: 42 }],
            ['function', 'action', { action() {} }],
            ['symbol', Symbol.iterator, { [Symbol.iterator]() {} }]
        ])('for a %s property', (type, key, source) => {
            const descriptors = {};
            const keys = [key];
            expect(getOwnPropertyDescriptors(source, descriptors, keys)).toBe(1);
            expect(descriptors).toMatchSnapshot();
        });
    });

    describe('does not list descriptors', () => {
        it('when the source is an empty object', () => {
            const source = {};
            const descriptors = {};
            const keys = ['foo'];
            expect(getOwnPropertyDescriptors(source, descriptors, keys)).toBe(0);
            expect(descriptors).toStrictEqual({});
        });

        it('when the list of keys is empty', () => {
            const source = { foo: 'bar' };
            const descriptors = {};
            const keys = [];
            expect(getOwnPropertyDescriptors(source, descriptors, keys)).toBe(0);
            expect(descriptors).toStrictEqual({});
        });

        it('when the expected properties does not exist in the source', () => {
            const source = { foo: 'bar' };
            const descriptors = {};
            const keys = ['bar'];
            expect(getOwnPropertyDescriptors(source, descriptors, keys)).toBe(0);
            expect(descriptors).toStrictEqual({});
        });

        it('when the properties already exist in the target', () => {
            const source = { foo: 'bar' };
            const target = { foo: 'baz' };
            const descriptors = {};
            const keys = ['foo'];
            expect(getOwnPropertyDescriptors(source, descriptors, keys, target)).toBe(0);
            expect(descriptors).toStrictEqual({});
        });

        it('when the properties are already listed', () => {
            const source = {};
            const target = { foo: 'bar' };
            const descriptors = { foo: 'baz' };
            const keys = ['foo'];
            expect(getOwnPropertyDescriptors(source, descriptors, keys, target)).toBe(0);
            expect(descriptors).toStrictEqual({ foo: 'baz' });
        });
    });
});
