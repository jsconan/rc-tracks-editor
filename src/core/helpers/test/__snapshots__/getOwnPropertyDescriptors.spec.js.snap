// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getOwnPropertyDescriptors lists descriptors for a boolean property 1`] = `
{
  "flag": {
    "configurable": true,
    "enumerable": true,
    "value": true,
    "writable": true,
  },
}
`;

exports[`getOwnPropertyDescriptors lists descriptors for a function property 1`] = `
{
  "action": {
    "configurable": true,
    "enumerable": true,
    "value": [Function],
    "writable": true,
  },
}
`;

exports[`getOwnPropertyDescriptors lists descriptors for a number property 1`] = `
{
  "answer": {
    "configurable": true,
    "enumerable": true,
    "value": 42,
    "writable": true,
  },
}
`;

exports[`getOwnPropertyDescriptors lists descriptors for a string property 1`] = `
{
  "message": {
    "configurable": true,
    "enumerable": true,
    "value": "hello",
    "writable": true,
  },
}
`;

exports[`getOwnPropertyDescriptors lists descriptors for a symbol property 1`] = `
{
  Symbol(Symbol.iterator): {
    "configurable": true,
    "enumerable": true,
    "value": [Function],
    "writable": true,
  },
}
`;
